### ISSUE TITLE
Needextract does not work with images

### ISSUE THREAD

**Issue by lehaas** [useblocks/sphinx-needs#818]
**Describe the bug**
We want to add images to the content of `need`s and reference these `need`s with `needextract`. Unfortunately, the images are not correctly parsed when extracting the need.

I assume it is related to https://github.com/useblocks/sphinx-needs/pull/794. However, the PR only solves the problem if images are located in `source/_images`. Where the images will also be located in `build`. 

**To Reproduce**
1. Create `projectA`:
   ```bash
   mkdir projectA
   cd projectA
   sphinx-quickstart 
   ```
   I choose to separate source and build folder during the build, i.e., `Separate source and build directories (y/n) [n]: y`
2. Configure `projectA` by adding the following lines to `projectA/source/conf.py`:
   ```python
   extensions = ["sphinx_needs"]
   ```
3. Add an image to the source folder, i.e.. `projectA/source/image.png`.
4. Create a requirement by adding the following lines in `projectA/source/index.rst`:
   ```rst
   .. need:: NeedA
      :id: NEEDA
   ```
5. Link the image with `needextract`:
   ```rst
   .. needextract:: NEEDA
   ```

**Expected behavior**
The requirement `NEEDA` appears twice and contains the used image twice.

**Actual behaviour**
For the first `need`, displays the image, the second `need` displays the path in `source/index.rst`
![image](https://user-images.githubusercontent.com/19656299/204226537-cdb9c87c-13ed-46ce-a62b-a21b0dc6e1cf.png)

Before using the most recent `sphinx-needs` release, the path was `_images/image.png`, i.e., the image copy in `build`. This broken as soon as the rst file was not on the `source` level but in a subfolder. 

**Setup**
We are using python Python `3.8.13`. And the following version of sphinx/sphinx-needs:
```bash
sphinx-needs                  1.1.0
Sphinx                        5.3.0
```

I hope the description helps to reproduce the issue :) 

The issue template was copied from [labs-zola](https://github.com/NYCPlanning/labs-zola).

**Comment by iSOLveIT** [comment]
Hi @lehaas,
First of all, I would like to thank you for the issue report.
I read through the description you gave on how to reproduce the issue and would like to clarify some things.

> We want to add images to the content of `needs` and reference these `needs` with `needextract`. Unfortunately, the images are not correctly parsed when extracting the need.
> I assume it is related to https://github.com/useblocks/sphinx-needs/pull/794. However, the PR only solves the problem if images are located in `source/_images`. Where the images will also be located in `build`.

The statement above is not entirely accurate because if you have an image directive (or an image) added to the content of a `need` directive, the images are copied and stored in the `_images` folder inside your `build` or `_build` directory.

**Example**

**Setup**
We are using python Python`3.9.13`. And the following version of sphinx/sphinx-needs:

```
sphinx-needs                  1.1.0
Sphinx                        5.3.0
```
**Project File Structure**

![image](https://user-images.githubusercontent.com/52050586/205237435-0529e97e-126d-42cf-a1f7-3247b4b56554.png)

1. Create `projectA`:
   ```bash
   mkdir projectA
   cd projectA
   sphinx-quickstart 
   ```
   I choose not to separate source and build folder during the build, i.e., `Separate source and build directories (y/n) [n]: n` but if you choose yes it should work.

2. Configure `projectA` by adding the following lines to `projectA/conf.py`:
   ```python
   extensions = ["sphinx_needs"]
   ```
3. Add an image to the project folder, i.e. `projectA/emoji.png`.
4. Create a requirement by adding the following lines in `projectA/index.rst`:
   ```rst
   .. need:: NeedA
      :id: NEEDA

      .. image:: emoji.png
   ```
5. Reference the `need` above with `needextract`:
   ```rst
   .. needextract:: NEEDA
   ```

**Expected behaviour**
The need `NEEDA` appears twice and contains the used image twice.

**Actual behaviour**
The need `NEEDA` appears twice and contains the used image twice.
![image](https://user-images.githubusercontent.com/52050586/205238344-7fff7982-71aa-444d-8d50-7409c2c1cad9.png)

**Comment by lehaas** [comment]
@iSOLveIT okay, maybe I broke down the example to far and it works now.. I was quite sure it did not work on my end. But you are right. That was my fault.

Let me give it another try, this time defining and linking the need in different files, as I actually want to use it and there it breaks on my end:

**To Reproduce**
Steps to reproduce the behavior:
1. Create `projectA`:
   ```bash
   mkdir projectA
   cd projectA
   sphinx-quickstart 
   ```
   I choose to separate source and build folder during the build, i.e., `Separate source and build directories (y/n) [n]: y`
2. Configure `projectA` by adding the following lines to `projectA/source/conf.py`:
   ```python
   extensions = ["sphinx_needs"]
   ```
3. Move to the `projectA/source`.
4. Add an image to the source folder, let us call it `image.png`.
5. Create a subfolder of `source` contain a file for the need definition `subfolder/image_need.rst` and add the following content :
 ```rst
Page
====

.. need:: NeedA
    :id: NEEDA

    .. image:: /image.png
 ```
6. Link the image with `needextract` in `index.rst:
```rst
.. toctree::
   :maxdepth: 2
   :caption: Contents:

   subfolder/image_need.rst

.. needextract:: NEEDA
```

**Expected behavior**
The requirement `NEEDA` appears in both files and each contains the used image.

**Actual behaviour**
For the first `need`, displays the image, the second `need` displays the path in `subfolder/image_need.html`
![grafik](https://user-images.githubusercontent.com/19656299/205255787-9ea9e039-c207-47ee-83c7-6d40f843ac9f.png)


In `index.html` the image link is broken: 
![grafik](https://user-images.githubusercontent.com/19656299/205255525-9a57ed82-26b2-4d51-bd97-05de63bf3cab.png)


**Setup**
We are using python Python `3.8.13` or `3.9.13`. And the following version of sphinx/sphinx-needs:
```bash
sphinx-needs                  1.1.0
Sphinx                        5.3.0
```

**Comment by iSOLveIT** [comment]
Hi @lehaas,

I read through the description you gave on how to reproduce the issue and here are my findings:
- In your dummy project, the images are not parsed as expected because the image link passed to the image directive is an absolute path and not a relative path. 
- Sphinx-Needs is working on making it possible to use absolute paths also. But you can use this solution to get the desired results. Instead of `.. image:: /image.png` you can use `.. image:: ../image.png` and the images will be parsed appropriately.

Here is the zipped file of my test project: [issue-818.zip](https://github.com/useblocks/sphinx-needs/files/10143935/issue-818.zip)

Thanks for the issue contribution.

**Comment by lehaas** [comment]
Hi @iSOLveIT ,

thanks for the answer. I am sad to hear that absolute paths don't work yet. Please let me know as soon as it is implemented.

On the relative paths: I don't see that relative paths solve the issue. In your code example, `US_002` was never used in a `need_extract`. If I add it to `index.rst`, again the image is not visualized correctly:
![grafik](https://user-images.githubusercontent.com/19656299/205579594-151c0efa-7513-4df8-8ea0-c97d736b4eb4.png)

Here is the zipped version: 
[issue-818.zip](https://github.com/useblocks/sphinx-needs/files/10151679/issue-818.zip)

Btw, I am using `make html` to build the project.

**Comment by danwos** [comment]
Hi all,

thanks for this detailed discussion. It helped me a lot to div deeper into the root cause of the problem.

First of all, the current implementation of needextract and images is nearly totally broken.

I have extended our test case for different scenarios, how images are referenced.
You can find it here: https://github.com/useblocks/sphinx-needs/tree/needextract_image/tests/doc_test/doc_needextract

The test-doc is creating needs incl. images with absolute and relative paths. 
Images are stored on root or in a subfolder. And the need is referenced by ``needextract`` in the same file or in a file outside need-folder.

Here are the results of what is working:
![image](https://user-images.githubusercontent.com/998700/205866754-2221967d-f3a7-4108-99c2-d84a98a3fd22.png)

As you see, only images stored inside ``/_images`` and referenced by a relative path are supported. Everything else does not work.

I see 2 possible error sources:

1. Sphinx-Needs is handling the image-node creation wrong. But as far as I know, we currently do nothing with paths and co. of referenced needs.
2 The Sphinx image handling is not triggered for the created image-nodes after `needextract` has copied the needs.

I'm afraid it is point 2, as Sphinx is doing the image-handling far before Sphinx-Needs is executing the `needextract` code. 
Sphinx image handling is directly executed when an image directive is detected. So in the `run()` function of the image directive.
Not sure, how we can reinvoke this behavior after `needextract` got executed.

**Comment by danwos** [comment]
Mhh the content get copied inside ``needextract`` via the `create_need()` function in `layout.py`.

https://github.com/useblocks/sphinx-needs/blob/56b4835ef6b3044e9bbde59681998bfa247fb549/sphinx_needs/layout.py#L47

So this line is reusing the complete content node-tree, which contains everything like text, table and maybe images-nodes.
It's using completely the same nodes as the original need, so I'm very confused why a reused image-node does not behave the same way.
Especially, why are relative paths working, but no absolute paths?

I also tried to create a copy via ` node_inner = needs[need_id]["content_node"].deepcopy()` instead of using the same referenced nodes.
But no change in the behavior.

**Comment by lehaas** [comment]
@danwos thanks a lot for the informations. We discovered one interesting point: Using `sphinx-needs==1.0.3`, images work just fine in `need_extract`. Not sure if there were any huge changes on your end that might have changed it. Maybe that can help you to ping down the error. 

>  Especially, why are relative paths working, but no absolute paths?
I don't agree with you that relative paths are working in `sphinx-needs==1.1.0`. On my end the table should as follows:
![grafik](https://user-images.githubusercontent.com/19656299/205909877-8600a442-5d99-4a3e-ade0-40597a688156.png)

Using relative paths, subfolder structure and storing the image in `_images` does not work on my end. See the following code:
[issue-818.zip](https://github.com/useblocks/sphinx-needs/files/10166290/issue-818.zip)

**Comment by danwos** [comment]
I make a quick check, if ``1.0.3`` behaves different as ``1.1.0``
And this is the result of `index.html`  under `tests/doc_test/doc_needextract`:
![image](https://user-images.githubusercontent.com/998700/205925401-83680388-9ac4-4a39-a606-75075764b5ee.png)

Now no `needextract` from a need inside `subfolder/file_1.rst` works. So my table and also the one from @lehaas do  not cover the new result. Can't say why.

These are the changes between `1.0.3` and `1.1.0`:
https://github.com/useblocks/sphinx-needs/compare/1.0.3...1.1.0

One of the major changes was, that we do not save any docutils nodes from the doctree anymore, as this would keep the complete doctree in memory.
Instead, we reference it by ID or create a copy, which is not part of the doctree anymore.
The last one happens also with the need-content node.
See [this line](https://github.com/useblocks/sphinx-needs/compare/1.0.3...1.1.0?diff=split#diff-a549d78f932228675000b620a8d55045057df704fbb3ddaa57ad0583beedd856R486)

Also the event-handling was reimplement a little different, but the order should be the same.
See [this line](https://github.com/useblocks/sphinx-needs/compare/1.0.3...1.1.0?diff=split#diff-81335d2905b404f2b2dffb86f60d7e1c0d537e457e1feafef7d9e46649464c0dR359)

I will revert some changes from `1.1.0` back to `1.0.3` to detect the source of the problem.

**Comment by arwedus** [comment]
@danwos: I noticed another problem with `needextract` in sphinx-needs 1.1.0.
It causes a sphinx error in our documentation project. I'll open another issue for it.

**Comment by lehaas** [comment]
@danwos I found another bug with images in `v1.0.3` and `v1.1.0`. If needs are referenced with `needextract` in multiple locations, the image link only works in the first reference in `v1.0.3`. The issue is further described in the PR. See https://github.com/useblocks/sphinx-needs/pull/837. 

Feel free to add it to your PR if you don't want to add a new folder.

**Comment by danwos** [comment]
Thanks all for testing and the detailed error reports.

I found a fix and integrated it already into PR #825.
Most important part is this one, which recalls Sphinx internal Transformators for need-content copied by ``needextract``.
https://github.com/useblocks/sphinx-needs/blob/869e6c2330a5482e0291a4ca57fdfe64dcfa3bb3/sphinx_needs/layout.py#L68-L71

This fixes also the multiple reference problem reported shortly by @lehaas.

Haven't found the time to clean the PR and finalize the tests, but will do it soon.

Any tests with your projects are welcome ;)
